src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java

```diff
@@ -262,97 +286,113 @@ private void _release(RevWalk walk) {
-                            @NonNull TaskListener listener)
+                            @NonNull final TaskListener listener)
             throws IOException, InterruptedException {
-        String cacheEntry = getCacheEntry();
-        Lock cacheLock = getCacheLock(cacheEntry);
-        cacheLock.lock();
-        try {
-            File cacheDir = getCacheDir(cacheEntry);
-            Git git = Git.with(listener, new EnvVars(EnvVars.masterEnvVars)).in(cacheDir);
-            GitTool tool = resolveGitTool();
-            if (tool != null) {
-                git.using(tool.getGitExe());
-            }
-            GitClient client = git.getClient();
-            client.addDefaultCredentials(getCredentials());
-            if (!client.hasGitRepo()) {
-                listener.getLogger().println("Creating git repository in " + cacheDir);
-                client.init();
-            }
-            String remoteName = getRemoteName();
-            listener.getLogger().println("Setting " + remoteName + " to " + getRemote());
-            client.setRemoteUrl(remoteName, getRemote());
-            listener.getLogger().println("Fetching & pruning " + remoteName + "...");
-            URIish remoteURI = null;
-            try {
-                remoteURI = new URIish(remoteName);
-            } catch (URISyntaxException ex) {
-                listener.getLogger().println("URI syntax exception for '" + remoteName + "' " + ex);
-            }
-            client.fetch_().prune().from(remoteURI, getRefSpecs()).execute();
-            final Repository repository = client.getRepository();
-            listener.getLogger().println("Getting remote branches...");
-            SCMSourceCriteria branchCriteria = getCriteria();
-            RevWalk walk = new RevWalk(repository);
-            try {
-                walk.setRetainBody(false);
-                for (Branch b : client.getRemoteBranches()) {
-                    if (!b.getName().startsWith(remoteName + "/")) {
-                      continue;
-                    }
-                    final String branchName = StringUtils.removeStart(b.getName(), remoteName + "/");
-                    listener.getLogger().println("Checking branch " + branchName);
-                    if (isExcluded(branchName)){
-                      continue;
-                    }
-                    if (branchCriteria != null) {
-                        RevCommit commit = walk.parseCommit(b.getSHA1());
-                        final long lastModified = TimeUnit.SECONDS.toMillis(commit.getCommitTime());
-                        final RevTree tree = commit.getTree();
-                        SCMSourceCriteria.Probe probe = new SCMSourceCriteria.Probe() {
-                            @Override
-                            public String name() {
-                                return branchName;
-                            }
+        doRetrieve(new Retriever<Void>() {
```

Comment:
Easier to follow when [ignoring indentation changes](https://github.com/jenkinsci/git-plugin/pull/433/files?w=1#diff-33b2f30e4119d2fd0e0bb8653c0c58bdR291).